planner/core/fragment.go

```diff
@@ -125,6 +127,41 @@ func partitionPruning(ctx sessionctx.Context, tbl table.PartitionedTable, conds
 // single physical table means a table without partitions or a single partition in a partition table.
 func (e *mppTaskGenerator) constructMPPTasksImpl(ctx context.Context, ts *PhysicalTableScan) ([]*kv.MPPTask, error) {
 	if ts != nil {
+		// update ranges according to correlated columns in access conditions like in the Open() of TableReaderExecutor
+		hasCorCol := false
+		for _, cond := range ts.AccessCondition {
+			if len(expression.ExtractCorColumns(cond)) > 0 {
+				hasCorCol = true
+				break
+			}
+		}
+		if hasCorCol {
+			access := ts.AccessCondition
+			if ts.Table.IsCommonHandle {
+				pkIdx := tables.FindPrimaryIndex(ts.Table)
+				idxCols, idxColLens := expression.IndexInfo2PrefixCols(ts.Columns, ts.Schema().Columns, pkIdx)
+				for _, cond := range access {
+					newCond, err := expression.SubstituteCorCol2Constant(cond)
+					if err != nil {
+						return nil, err
+					}
+					access = append(access, newCond)
+				}
+				res, err := ranger.DetachCondAndBuildRangeForIndex(e.ctx, access, idxCols, idxColLens)
+				if err != nil {
+					return nil, err
+				}
+				ts.Ranges = res.Ranges
+			} else {
+				var err error
+				pkTP := ts.Table.GetPkColInfo().FieldType
+				ts.Ranges, err = ranger.BuildTableRange(access, e.ctx.GetSessionVars().StmtCtx, &pkTP)
+				if err != nil {
+					return nil, err
+				}
+			}
```

Comment:
Could you please extract this logic to a separate function `handleCorcols(talbeInfo, conds, ...)` so that both `mppTaskGenerator` and `TableReaderExecutor` can use it?

Reply:
fixed.